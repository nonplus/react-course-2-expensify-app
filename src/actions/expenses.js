import uuid from "uuid";
import database from "../firebase/firebase";
import _ from "lodash";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const addExpenseAsync = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const ref = await database.ref("expenses").push(expense);
    await dispatch(addExpense({ id: ref.key, ...expense }));
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const setExpenseAsync = () => {
  return async dispatch => {
    const snapshot = await database.ref("expenses").once("value");
    const expenses = _.map(snapshot.val(), (value, id) => ({ id, ...value }));
    dispatch(setExpenses(expenses));
  };
};

export const editExpense = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return {
    type: "EDIT_EXPENSE",
    id,
    updates: {
      description,
      note,
      amount,
      createdAt
    }
  };
};

export const editExpenseAsync = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return async dispatch => {
    await database
      .ref(`expenses/${id}`)
      .set({ description, note, amount, createdAt });
    dispatch(editExpense(id, updates));
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const removeExpenseAsync = ({ id } = {}) => {
  return async dispatch => {
    await database.ref(`expenses/${id}`).remove();
    dispatch(removeExpense({ id }));
  };
};
