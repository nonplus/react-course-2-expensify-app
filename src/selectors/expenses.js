export const selectExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  const startDateMatches = startDate
    ? expense => startDate.isSameOrBefore(expense.createdAt, "day")
    : () => true;
  const endDateMatches = endDate
    ? expense => endDate.isSameOrAfter(expense.createdAt, "day")
    : () => true;
  text = text.toLowerCase();
  const textMatches = !text
    ? () => true
    : expense => expense.description.toLowerCase().includes(text);

  const sortCompare =
    sortBy === "date"
      ? (a, b) => b.createdAt - a.createdAt
      : sortBy === "amount" ? (a, b) => b.amount - a.amount : () => 0;

  return expenses
    .filter(
      expense =>
        startDateMatches(expense) &&
        endDateMatches(expense) &&
        textMatches(expense)
    )
    .sort(sortCompare);
};
