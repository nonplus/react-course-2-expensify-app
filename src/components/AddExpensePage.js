import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpenseAsync } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
  }

  onSubmit = async expense => {
    await this.props.addExpenseAsync(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 class="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapPropsToDispatch = dispatch => ({
  addExpenseAsync: expense => dispatch(addExpenseAsync(expense))
});

export default connect(undefined, mapPropsToDispatch)(AddExpensePage);
