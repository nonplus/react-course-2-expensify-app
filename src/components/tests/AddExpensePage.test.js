import React from "react";
import { shallow } from "enzyme";
import expenses from "../../tests/fixtures/expenses";
import { AddExpensePage } from "../AddExpensePage";

describe("<AddExpensePage />", () => {
  let addExpenseAsync, history, wrapper;

  beforeEach(() => {
    addExpenseAsync = jest.fn();
    history = {
      push: jest.fn()
    };
    wrapper = shallow(
      <AddExpensePage addExpenseAsync={addExpenseAsync} history={history} />
    );
  });

  it("should render the page correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("onSubmit", () => {
    let addExpenseAsyncPromise;
    beforeEach(() => {
      addExpenseAsyncPromise = Promise.resolve();
      addExpenseAsync.mockImplementation(() => addExpenseAsyncPromise);
      wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
    });

    it("should call props.addExpenseAsync(expense)", () => {
      expect(addExpenseAsync).toHaveBeenLastCalledWith(expenses[1]);
    });
    it("should call history.push('/')", async () => {
      await addExpenseAsyncPromise;
      expect(history.push).toHaveBeenLastCalledWith("/");
    });
  });
});
