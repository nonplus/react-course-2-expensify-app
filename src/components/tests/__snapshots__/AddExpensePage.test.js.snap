// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddExpensePage /> should render the page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        class="page-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
